satellite/metainfo/loop.go

```diff
@@ -0,0 +1,214 @@
+// Copyright (C) 2019 Storj Labs, Inc.
+// See LICENSE for copying information.
+
+package metainfo
+
+import (
+	"context"
+	"time"
+
+	"github.com/gogo/protobuf/proto"
+	"github.com/zeebo/errs"
+
+	"storj.io/storj/pkg/pb"
+	"storj.io/storj/pkg/storj"
+	"storj.io/storj/storage"
+)
+
+var (
+	// LoopError is a standard error class for this component.
+	LoopError = errs.Class("metainfo loop error")
+)
+
+// Observer is an interface defining an observer that can subscribe to the metainfo loop.
+type Observer interface {
+	RemoteSegment(context.Context, storj.Path, *pb.Pointer) error
+	RemoteObject(context.Context, storj.Path, *pb.Pointer) error
+	InlineSegment(context.Context, storj.Path, *pb.Pointer) error
+}
+
+type observerContext struct {
+	Observer
+	ctx  context.Context
+	done chan error
+}
+
+func (observer *observerContext) HandleError(err error) bool {
+	if err != nil {
+		observer.done <- err
+		observer.Finish()
+		return true
+	}
+	return false
+}
+
+func (observer *observerContext) Finish() {
+	close(observer.done)
+}
+
+func (observer *observerContext) Wait() error {
+	return <-observer.done
+}
+
+// LoopConfig contains configurable values for the metainfo loop.
+type LoopConfig struct {
+	CoalesceDuration time.Duration `help:"how long to wait for new observers before starting iteration" releaseDefault:"5s" devDefault:"5s"`
+}
+
+// Loop is a metainfo loop service.
+type Loop struct {
+	config   LoopConfig
+	metainfo *Service
+	join     chan *observerContext
+	done     chan struct{}
+}
+
+// NewLoop creates a new metainfo loop service.
+func NewLoop(config LoopConfig, metainfo *Service) *Loop {
+	return &Loop{
+		metainfo: metainfo,
+		config:   config,
+		join:     make(chan *observerContext),
+		done:     make(chan struct{}),
+	}
+}
+
+// Run starts the looping service.
+func (loop *Loop) Run(ctx context.Context) (err error) {
+	defer mon.Task()(&ctx)(&err)
+
+	defer close(loop.done)
+
+	for {
+		err := loop.runOnce(ctx)
+		if err != nil {
+			return err
+		}
+	}
+}
+
+// Join will join the looper for one full cycle until completion and then returns.
+// On ctx cancel the observer will return without completely finishing.
+// Only on full complete iteration it will return nil.
+func (loop *Loop) Join(ctx context.Context, observer Observer) (err error) {
+	defer mon.Task()(&ctx)(&err)
+
+	obsContext := &observerContext{
+		Observer: observer,
+		ctx:      ctx,
+		done:     make(chan error),
+	}
+
+	select {
+	case <-loop.done:
+		return LoopError.New("loop closed")
```

Comment:
I think that it will be preferable to be able to identify this error over others in the way that you the caller can inspect it using the [`Is` function](https://godoc.org/github.com/zeebo/errs#Is), however it would depend in how `Close` gets implemented.